docs/design/event_compression.md

```diff
@@ -0,0 +1,79 @@
+
+Event compression should be best effort (not guaranteed). Meaning, in the worst case, ```n``` identical (minus timestamp) events may still result in ```n``` event entries.
+
+## Design
+Instead of a single Timestamp, each event object [contains](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/types.go#L1111) the following fields:
+ * ```FirstTimestamp util.Time``` 
+   * The date/time of the first occurrence of the event.
+ * ```LastTimestamp util.Time```
+   * The date/time of the most recent occurrence of the event.
+   * On first occurrence, this is equal to the FirstTimestamp.
+ * ```Count int```
+   * The number of occurrences of this event between FirstTimestamp and LastTimestamp
+   * On first occurrence, this is 1.
+
+Each binary that generates events will:
+ * Maintain a new global hash table to keep track of previously generated events (see ```pkg/client/record/events_cache.go```).
+ * The code that “records/writes” events (see ```StartRecording``` in ```pkg/client/record/event.go```), uses the global hash table to check if any new event has been seen previously.
+   * The key for the hash table is generated from the event object minus timestamps/count/transient fields (see ```pkg/client/record/events_cache.go```), specifically the following events fields are used to construct a unique key for an event:
+     * ```event.Source.Component```
+     * ```event.Source.Host```
+     * ```event.InvolvedObject.Kind```
+     * ```event.InvolvedObject.Namespace```
+     * ```event.InvolvedObject.Name```
+     * ```event.InvolvedObject.UID```
+     * ```event.InvolvedObject.APIVersion```
+     * ```event.Reason```
+     * ```event.Message```
+   * If the key for a new event matches the key for a previously generated events (meaning all of the above fields match between the new event and some previously generated event), then the event is considered to be a duplicate:
+     * Instead of the usual POST/create event API, the new PUT (update) event API is called to update the existing event entry in etcd with the new last seen timestamp and count.
+     * The event is also updated in the global hash table with an incremented count, updated last seen timestamp, name, and new resource version (all required to issue a future event update).
+   * If the key for a new event does not match the key for any previously generated event (meaning none of the above fields match between the new event and any previously generated events), then the event is considered to be new/unique:
+     * The usual POST/create event API is called to create a new event entry in etcd.
+     * An entry for the event is also added to the global hash table.
+
+## Issues/Risks
+ * Hash table clean up
+   * If the component (e.g. kubelet) runs for a long period of time and generates a ton of unique events, the hash table could grow very large in memory.
+   * *Future consideration:* remove entries from the hash table that are older than some specified time.
```

Comment:
This should be handled now, not in the future.
The cache should be fixed size, not a time-based cleanup.
Consider using an lru cache like https://github.com/hashicorp/golang-lru.